spec/blocksync/verification.md

```diff
@@ -0,0 +1,120 @@
+
+**Switching from statesync to blocksync**
+
+If we switch from statesync to blocksync, the last item in the block store is not the entire block. The store holds only the verified header and the commit that verified it. Therefore, we do need to fetch the corresponding block from our peers, verify it against data stored in the store and set the `last trusted block` to this block. 
+
+To enable this, upon switching from statesync, we do not instruct the block pool to fetch blocks at height `H + 1` (where height `H` corresponds to the height of the header last verified with statesync). Instead we let it fetch the block at height H, but do not execute this block against the application (as this would fail - state is already updated to height H + 1. )
+
+### Verifying blocks past the trusted state
+
+<img src="img/bc-reactor-blocks.png" alt="block diagram" title="Block overview" width="80%" name="Block_overview"/>
+
+
+The diagram above shows all blocks at play when verifying block at height ` H + 1` - `newBlock`, where at height `H` we have the trusted block. 
+
+At a high level, the verification algorithm is represented by the code below. 
+
+~~~golang
+if LastCommit.verify(trustedBlock) {
+        // NewBlock is invalid, request new
+            request_newBlock_from_other_peer(height: h + 1)
+} else { 
+    if NewBlock.validatorSet == trustedState.validatorSet {		     
+        if  LastCommit2.validatorSetHash == trustedState.validatorSetHash && 
+		verifyBlock.validatorSetHash == trustedState.nextValidatorSetHash {
+			if LastCommit2.verify(NewBlock) {
+                        if verify_block_with_peers(NewBlock) {
+								ApplyBlock(NewBlock) 
+                        }
+            }|
+        } else
+			find_new_verifyBlock(height: h + 2)
+    }
+}
+
+~~~
+
+We try to increase the trust in the NewBlock by first confirming that  it's `lastCommit` field indeed verifies the trusted block. We then confirm that the validator hash matches the expected hash. But to fully verify the block we need to check it against the signatures that signed it. They are stored within the `lastCommit2` of `verifyBlock` at height `H+2`. As we have no gurantees on this block, we again increase our trust in it by confirming that the validators whose signatures are in `lastCommit2` are a match for the expected validator set.
+
+#### *Witness verification*
+
+**Note**. This is not implemented yet. 
+
+If all these checks pass, the reactor verifies that other peers have the same block at the particular height. 
+
+In order to reduce traffic, we do not ask the peers to provide the whole block to us, rather only the header. If crosschecking the headers fails then the node requests the remainder of the block to decide on whether this peer is faulty or not.
```

Comment:
To further reduce traffic, we could just request the block's hash (smaller than the header)